include/range/v3/action/slice.hpp

```diff
@@ -36,47 +36,100 @@ namespace ranges
+              static auto bind(slice_fn slice, D from, D to)
+              RANGES_DECLTYPE_AUTO_RETURN
+              (
+                  std::bind(slice, std::placeholders::_1, from, to)
+              )
+              template<typename D, CONCEPT_REQUIRES_(Integral<D>())>
+              static auto bind(slice_fn slice, D from, detail::from_end_<D> to)
+              RANGES_DECLTYPE_AUTO_RETURN
+              (
+                  std::bind(slice, std::placeholders::_1, from, to)
+              )
+              template<typename D, CONCEPT_REQUIRES_(Integral<D>())>
+              static auto bind(slice_fn slice, detail::from_end_<D> from, detail::from_end_<D> to)
+              RANGES_DECLTYPE_AUTO_RETURN
+              (
+                  std::bind(slice, std::placeholders::_1, from, to)
+              )
+              template<typename D, CONCEPT_REQUIRES_(Integral<D>())>
+              static auto bind(slice_fn slice, D from, const end_fn &to)
+              RANGES_DECLTYPE_AUTO_RETURN
+              (
+                  std::bind(slice, std::placeholders::_1, from, to)
+              )
+              template<typename D, CONCEPT_REQUIRES_(Integral<D>())>
+              static auto bind(slice_fn slice, detail::from_end_<D> from, const end_fn &to)
+              RANGES_DECLTYPE_AUTO_RETURN
+              (
+                  std::bind(slice, std::placeholders::_1, from, to)
+              )
+            public:
 
-                // TODO support slice from end.
-                template<typename Rng,
-                    typename I = iterator_t<Rng>,
-                    typename D = range_difference_type_t<Rng>,
-                    CONCEPT_REQUIRES_(Concept<Rng, D, D>())>
-                Rng operator()(Rng && rng, range_difference_type_t<Rng> from,
-                    range_difference_type_t<Rng> to) const
+              template<typename Rng,
+                  typename I = iterator_t<Rng>,
+                  typename D = range_difference_type_t<Rng>,
+                  CONCEPT_REQUIRES_(ForwardRange<Rng>() && ErasableRange<Rng, I, I>())>
+              Rng operator()(Rng &&rng, D from, D to) const
                 {
-                    RANGES_EXPECT(from <= to);
+                    RANGES_EXPECT(0 <= from && 0 <= to && from <= to);
                     ranges::action::erase(rng, next(begin(rng), to), end(rng));
                     ranges::action::erase(rng, begin(rng), next(begin(rng), from));
-                    return static_cast<Rng&&>(rng);
+                    return static_cast<Rng &&>(rng);
                 }
 
+              template<typename Rng,
+                  typename I = iterator_t<Rng>,
+                  typename D = range_difference_type_t<Rng>,
+                  CONCEPT_REQUIRES_(BidirectionalRange<Rng>() && ErasableRange<Rng, I, I>())>
+              Rng operator()(Rng &&rng, D from, detail::from_end_<D> to) const {
+                  RANGES_EXPECT(0 <= from && to.dist_ <= 0);
```

Comment:
`RANGES_EXPECT(!SizedRange<Rng> || from - to.dist_ <= distance(rng));` wouldn't hurt here. Or even the potentially O(N) `RANGES_EXPECT(from - to.dist_ <= distance(rng));` since we're about to perform O(N) erasures anyway.

Reply:
Im leaning towards keeping the `!SizedRange` since i do not want the extra overhead. But i can change it if you prefer it

Reply:
I have no strong preference either way. (I have some vague icky feelings about `slice` not always being O(1), but that's a pre-existing property and not specific to this change.)